File: renderbot.py
import json
import argparse
import sys
import os
import asyncio
import time
import traceback

parser = argparse.ArgumentParser(description="Maya render automation")
parser.add_argument("config", nargs="?", help="Configuration .json file", default="renderbot.json")
parser.add_argument("--root", nargs="?", help="Root directory")
argv = parser.parse_args(sys.argv[1:])

with open(argv.config) as f:
    config = json.load(f)

def merge_dict(x, y):
    z = x.copy()
    z.update(y)
    return z

def log_bytes(b):
    if b: print(b.encode("utf-8"))

async def render_communicate(proc, take, config):
    name = take["name"]

    last_ping = time.time()
    code = -99999
    while True:
        try:
            code = await asyncio.wait_for(proc.wait(), timeout=0.1)
            break
        except asyncio.TimeoutError:
            pass
        except asyncio.CancelledError:
            pass

        if (time.time() - last_ping) > 60*10:
            break

        try:
            for n in range(20):
                line = await asyncio.wait_for(proc.stdout.readline(), timeout=5)
                line = line.strip()
                if line:
                    last_ping = time.time()
                    print(line.decode("utf-8", errors="ignore"), flush=True)
                if (time.time() - last_ping) > 5: break
        except asyncio.TimeoutError:
            if (time.time() - last_ping) > 30:
                print("No response.. trying to send CRLF", flush=True)
                try:
                    await asyncio.wait_for(proc.stdin.write(b"\r\n"), timeout=5)
                except asyncio.TimeoutError:
                    print("Failed to send CRLF...", flush=True)
        except asyncio.CancelledError:
            pass
    
    print("")
    print(f"=== Finished render '{name}' ===")
    print("")
    if code == -99999:
        print("Timeout... Renderer probably hanged")
    else:
        if code == 0:
            print("Success!")
        else:
            print(f"Error code: {code}")
    return code

async def render_take(take, config):
    take = merge_dict(config.get("defaults", {}), take)
    render_exe = config["render-exe"]
    root = config.get("root-dir", ".")
    if argv.root: root = argv.root
    name, filename = take["name"], take["file"]
    start, end = take["start"], take["end"]
    step = take.get("step", 1)

    args = [render_exe]

    # Frame range
    args += ["-s", str(start)]
    args += ["-e", str(end)]
    args += ["-b", str(step)]

    # User arguments
    args += ["-r", take.get("renderer", "arnold")]
    args += ["-rd", take.get("output", ".")]
    args += ["-skipExistingFrames", ["false", "true"][take.get("skipExisting", False)]]

    # Finish with filename
    args.append(os.path.join(root, filename))

    print("")
    print(f"=== Starting render '{name}' ===")
    print("")
    print("$ " + " ".join(args), flush=True)
    print("")

    proc = await asyncio.subprocess.create_subprocess_exec(*args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.STDOUT,
        cwd=root,
    )

    return await render_communicate(proc, take, config)

for take in config["takes"]:
    for n in range(10):
        try:
            code = asyncio.run(render_take(take, config))
            if code == 0:
                break
        except:
            print("")
            print("=== renderbot.py internal error ===")
            print("")
            traceback.print_exc()

File: renderbot.js
const path = require("path");
const { performance } = require("perf_hooks");
const fsp = require("fs").promises;
const { spawn } = require("child_process");
const { fail } = require("assert");

function parseArgs(argv) {
    const args = { };
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i];
        if (arg.startsWith("--")) {
            if (++i >= process.argv) fail(`Expected value for '${arg}'`);
            args[arg.substring(2)] = argv[i];
        } else {
            args.configFiles = [...(args.configFiles || []), arg];
        }
    }
    return args;
}

function humanTimeMs(floatMs) {
    const ms = floatMs | 0;
    const s = ms / 1000 | 0;
    const m = s / 60 | 0;
    const h = m / 60 | 0;
    if (h > 0) {
        return `${h}h ${m%60}m`;
    } else if (m > 0) {
        return `${m}m ${s%60}s`;
    } else if (s > 0) {
        return `${s}s`;
    } else {
        return `${ms}ms`;
    }
}

function humanTimeS(floatSec) {
    return humanTimeMs(floatSec * 1000);
}

async function renderTake(args, config, take) {
    const {
        name = fail("Missing 'name': User-defined name for the take"),
        file = fail("Missing 'file': Maya scene .ma or .mb file to render"),
        renderExe = fail("Missing 'renderExe': path pointing to Maya Render.exe"),
        start = fail("Missing 'start': first frame to render"),
        end = fail("Missing 'end': last frame to render"),
        step = 1,
        rootDir = ".",
        outputDir = ".",
        renderer = "arnold",
        skipExistingFrames = false,
        pingWait = 30,
        pingInterval = 2,
        timeout = 15*60,
    } = { ...config, ...take, ...args };
    const renderArgs = {
        ...config.options,
        ...take.options,
        ...args.options,
    };

    const argv = [];

    // Output options
    argv.push("-rd", outputDir)
    argv.push("-r", renderer)

    // Frame range
    argv.push("-s", start)
    argv.push("-e", end)
    argv.push("-b", step)

    // Custom arguments
    for (const [key, value] of Object.entries(renderArgs)) {
        argv.push(`-${key}`, value.toString());
    }

    // Filename
    argv.push(file);

    console.log(`$ ${renderExe} ${argv.join(" ")}\n`);

    const proc = spawn(renderExe, argv, {
        cwd: rootDir,
    });

    let lastOutput = performance.now();

    // Forward stdout and stderr and keep track of last output
    proc.stdout.on("data", (data) => {
        lastOutput = performance.now();
        process.stdout.write(data);
    });
    proc.stderr.on("data", (data) => {
        lastOutput = performance.now();
        process.stderr.write(data);
    });

    // Spam CRLF to stdin if we haven't received any output in a while
    let shouldPrintCrlfInfo = true;
    const crlfSpamInterval = setInterval(() => {
        const ping = performance.now() - lastOutput;
        if (ping > pingWait*1000) {
            if (shouldPrintCrlfInfo) {
                console.log(`No input for ${humanTimeS(pingWait)}.. sending CRLF every ${humanTimeS(pingInterval)}s`);
                shouldPrintCrlfInfo = false;
            }
            try {
                proc.stdin.write("\r\n");
            } catch (error) {
                console.error("Sending CRLF failed..", error);
            }
        } else {
            shouldPrintCrlfInfo = true;
        }
    }, pingInterval*1000);

    // Terminate the program if we haven't received any output for a _long_ while
    const timeoutInterval = setInterval(() => {
        const ping = performance.now() - lastOutput;
        if (ping > timeout*1000) {
            console.log(`No input for ${humanTimeS(ping)}.. Assuming hand and attempting to terminate`);
            try {
                proc.kill();
            } catch (error) {
                console.error("Failed to terminate", error);
            }
        }
    }, 10000);

    await new Promise((resolve, reject) => {
        proc.on("exit", (code) => {
            clearInterval(crlfSpamInterval);
            clearInterval(timeoutInterval);
            if (code == 0) {
                resolve();
            } else {
                reject(new Error(`Process failed with code ${code}`));
            }
        });
    });
}

async function renderConfig(args, configFile) {
    const configData = await fsp.readFile(configFile, {
        encoding: "utf-8",
    })
    const config = JSON.parse(configData);

    const {
        takes = []
    } = { ...config, ...args };

    for (const take of takes) {
        const {
            name = fail("Missing 'name': User-defined name for the take"),
            restarts = 5,
        } = { ...config, ...take, ...args };

        console.log(`\n=== Starting take '${take.name}' ===\n`);

        const begin = performance.now();

        for (let restartI = 0; restartI <= restarts; restartI++) {
            try {
                await renderTake(args, config, take);
            } catch (error) {
                console.error("");
                console.error(error);
                console.error(`\n=== Take failed '${take.name}' ===\n`);
                if (restartI + 1 <= restarts) {
                    console.error(`... restarting ${restartI+1}/${restarts}\n`);
                }
            }
        }

        const end = performance.now();

        console.log(`\n=== Finished take '${take.name}' in ${humanTimeMs(end - begin)} ===\n`);
    }
}

async function main(args) {
    const {
        configFiles = ["renderbot.json"],
    } = args;

    console.log("\nUsing arguments:", args);
    for (const configFile of configFiles) {
        console.log(`\n=== Starting config '${configFile}' ===\n`);

        const begin = performance.now();
        await renderConfig(args, configFile);
        const end = performance.now();

        console.log(`\n=== Finished config '${configFile}' in ${humanTimeMs(end - begin)} ===\n`);
    }

}

const argv = process.argv.slice(2);
const args = parseArgs(argv);
main(args);
